design-proposals/descheduler-support.md

```diff
@@ -0,0 +1,105 @@
+
+
+### Disruption Budget Controller
+`virt-launcher` pods have another layer of protection against evictions using [PDBs](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#pod-disruption-budgets).
+
+The controller is responsible for managing PDBs:
+
+The controller decides whether a VMI needs protection against eviction based on its migration strategy (defined on the VMI with a fallback to a cluster-wide definition):
+
+| Migration Strategy    | Requires a PDB                 | 
+|-----------------------|--------------------------------|
+| None                  | False                          |
+| LiveMigrate           | True                           |
+| LiveMigrateIfPossible | Only if the VMI is migrateable |
+| External              | True                           |
+
+### Evacuation Controller
+The controller looks for potential VMIs to evict and tries to migrate them to another node.
+
+### Migration Controller
+The controller manages the migration process initiated by the Evacuation controller.
+The controller migrates the `virt-launcher` pod from the source node to a target node.
+After the migration process is completed, the original `virt-launcher` pod is deleted from the source node and the eviction process ends.
+
+## Motivation
+Cluster administrators want their workloads to be balanced across the cluster.
+These workloads include KubeVirt VMs.
+
+## Goals
+1. Enable KubeVirt to work on existing K8s clusters that already have a descheduler deployed.
+2. Enable cluster admins to introduce a descheduler to an existing cluster with KubeVirt deployed.
+
+## Non Goals
+The descheduler should respect the two annotations included in this document.
+
+## Definition of Users
+- Cluster administrator
+
+## User Stories
+- As a cluster admin running a K8s cluster with a descheduler, I want to deploy and use KubeVirt.
+- As a cluster admin running a K8s cluster with KubeVirt, I want to deploy a descheduler to load balance all my workloads.
+- As a cluster admin I want to enable cluster-wide descheduler support.
+
+## Repos
+1. kubevirt/kubevirt
+2. kubevirt/hyperconverged-cluster-operator
+
+# Design
+
+1. A `DeschedulerSupport` feature gate will be introduced in order to enable this feature.
+2. The feature gate will be disabled by default.
+3. The VMI controller will add the `descheduler.alpha.kubernetes.io/request-evict-only` to newly created pods, or to existing pods that do not have them.
+4. The VMI controller will add the `descheduler.alpha.kubernetes.io/eviction-in-progress` annotation to the `virt-launcher` pod, if the VMI is marked for eviction.
+
+## API Examples
+(tangible API examples used for discussion)
+
+## Scalability
+This feature should not affect scalability.
+
+## Update/Rollback Compatibility
+The VMI-controller reconciles all `virt-launcher` pods, so it will annotate existing `virt-launcher` pods as well.
```

Comment:
To avoid confusion, I suggest mentioning that the VMI-controller is part of the virt-controller operator.

Reply:
Done.